etcdserver/corrupt.go

```diff
@@ -19,11 +19,65 @@ import (
 	"time"
 
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 	"github.com/coreos/etcd/mvcc"
 	"github.com/coreos/etcd/pkg/types"
 )
 
+// checkInitialHashKV compares initial hash values with its peers
+// before serving any client traffic.
+func (s *EtcdServer) checkInitialHashKV() {
+	if !s.Cfg.InitialCorruptCheck {
+		return
+	}
+
+	// wait until server is ready by publishing attributes
+	// to "/0/members/*", so that *membership.RaftCluster
+	// is up-to-date with its peers, on restart
+	select {
+	case <-s.readych:
+	case <-s.stopping:
+		return
+	}
+
+	plog.Infof("%s starting initial corruption check...", s.ID())
+	h, rev, crev, err := s.kv.HashByRev(0)
+	if err != nil {
+		plog.Fatalf("%s failed to fetch hash (%v)", s.ID(), err)
+	}
+
+	peers := s.getPeerHashKVs(rev)
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp == nil {
+			if p.err != nil {
+				if p.err == rpctypes.ErrFutureRev {
+					// e.g. db file is corrupted so that hash-kv was requested with future revision
+					// then peer will return "mvcc: required revision is a future revision"
+					mismatch++
+				}
+				// if this local node is slow (e.g. ErrCompacted) on requested revision,
+				// then it's not corrupt; just this node is slow
+			}
+			continue
+		}
+		peerID := types.ID(p.resp.Header.MemberId)
+		if h != p.resp.Hash && rev == p.resp.Header.Revision && crev == p.resp.CompactRevision {
```

Comment:
why do we need to compare rev and crev? we request hash at rev. it if fails, error should be reported. we should trust the server response.